### PR TITLE
Add catalog indexer worker

### LINKED ISSUE #4147: Implement new catalog indexer-worker
## Problem

This issue tracks the creation of a new catalog-indexer-worker Docker image. It does not include adding any orchestration steps to the DAG, or any infrastructure work to actually create ASGs.

## Description

First we will create a new indexer-worker directory under catalog/dags/data_refresh, which will contain the contents of the new indexer worker. This implementation already exists in the ingestion server. The relevant pieces can be pulled out and refactored slightly to fit the new, much smaller image. Broadly, this is the mapping of existing files to new files needed:

- `api.py` will defined the API for the worker, and is refactored from the existing [indexer_worker.py](https://github.com/WordPress/openverse/blob/05ff48d05f2163104151c5589cf352a156bc6a97/ingestion_server/ingestion_server/indexer_worker.py). It must be refactored to add task state and a task_status endpoint, which takes a task_id and returns the status and progress of the given task.
- `indexer.py` will contain the logic for the actual indexing task. It will be refactored from the existing [indexer.py](https://github.com/WordPress/openverse/blob/05ff48d05f2163104151c5589cf352a156bc6a97/ingestion_server/ingestion_server/indexer.py); specifically all we need is the [replicate](https://github.com/WordPress/openverse/blob/05ff48d05f2163104151c5589cf352a156bc6a97/ingestion_server/ingestion_server/indexer.py#L157) function.
- `elasticsearch_models.py`, pulled from the file of the [same name](https://github.com/WordPress/openverse/blob/05ff48d05f2163104151c5589cf352a156bc6a97/ingestion_server/ingestion_server/elasticsearch_models.py) in the ingestion server. Defines a mapping from a database record to an Elasticsearch document.
- Utility files for helper functions for connecting to Elasticsearch and Postgres (e.g. [es_helpers.py](https://github.com/WordPress/openverse/blob/main/ingestion_server/ingestion_server/es_helpers.py))

The [Dockerfile](https://github.com/WordPress/openverse/blob/main/ingestion_server/Dockerfile) can be copied from the existing ingestion server. It should be updated to reference the new file structure, and to expose only a single port, which should be distinguished from the ports currently in use by the ingestion server (8001 and 8002). Other necessary files, including env.docker, .dockerignore, Pipfile, and gunicorn.conf.py can all be copied in from the existing ingestion server as well.

Finally we will update the monorepo’s root [docker-compose.yml](https://github.com/WordPress/openverse/blob/05ff48d05f2163104151c5589cf352a156bc6a97/docker-compose.yml) to add a new catalog-indexer-worker service. Its build context should point to the nested data_refresh/indexer_worker directory, and it should map the exposed port to enable the API to be reached by the catalog.

When this work is complete, it should be possible to run `just catalog/shell` and curl the new indexer worker. The existing ingestion-server and indexer-worker services are unaffected (it is still possible to run legacy data refreshes locally and in production).

## Additional context

See [this section](https://docs.openverse.org/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.html#implement-new-catalog-indexer-worker) of the IP.

### PR DESCRIPTION
**Pull request by stacimc**
## Fixes

Fixes #4147 by @stacimc 

## Description

This PR adds the catalog indexer worker. It does not actually connect to it in the data refresh yet.

Some unit tests are added, but because of the nature of the worker the bulk of the tests will be integration tests modeled after the existing ingestion server integration tests. I was able to get these mostly working locally but pulled them out of this PR as I think it will make more sense to add the integration tests once the data refresh is fully implemented, possibly reusing issue #4340.

Almost all of this code is just a refactor of existing code. Where possible I've noted when files are almost entirely copied over from the ingestion server. 

Note that the service is called `catalog_indexer_worker` to distinguish it from the existing `indexer_worker` that belongs to the ingestion server, which we must keep around for now.

## Testing Instructions

`just build && just up`

Use [elasticvue](https://app.elasticvue.com/cluster/0/indices) to `Delete All Documents` from one of your existing indices. I used `audio-init-filtered`. The index should now have 0 documents.

Now run `just catalog/shell`, and you should be able to curl the indexer worker.

```
# Replace the target_index name if you don't have audio-init-filtered
# Note the start_id/end_id we're using, to make it reindex only 100 records
curl -X POST -d '{"model_name": "audio", "table_name": "audio", "start_id": 0, "end_id": 100, "target_index": "audio-init-filtered"}' -H "Content-Type: application/json" 'http://catalog_indexer_worker:8003/task'
```

You should get back a response like:
```
{
  "message": "Successfully scheduled task.", 
  "task_id": "adbec2cbc93a4c2f895caf12ede3eb88", 
  "status_check": "http://catalog_indexer_worker:8003/task/adbec2cbc93a4c2f895caf12ede3eb88"
}
```

Now curl the status_check endpoint that was just returned:
```
curl 'http://catalog_indexer_worker:8003/task/adbec2cbc93a4c2f895caf12ede3eb88'
```

And you should see something like:

```
{
  "task_id": "adbec2cbc93a4c2f895caf12ede3eb88",
  "active": false,
  "model": "audio",
  "target_index": "audio-init-filtered",
  "progress": 100.0,
  "start_time": "2024-05-14 21:27:54.437745",
  "finish_time": "2024-05-14 21:27:54.575137",
  "error": false
}
```

In elasticvue, you should now see 100 docs in your target index.

If you used a filtered index, you can now query the API (http://localhost:50281/v1/audio) and see that the results are returned as expected. **⚠️ Important: In Elasticvue you must first refresh the index to make the documents available. Then you can inspect them in Elasticvue, or check the API. This will be handled by the DAG later.**

You can also try running the `just` commands for the indexer worker.

## Checklist

- [ ] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [ ] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (`just catalog/generate-docs` for catalog
      PRs) or the media properties generator (`just catalog/generate-docs media-props`
      for the catalog or `just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: catalog/dags/data_refresh/indexer_worker/indexer_worker/authority.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+"""
```
Comment (stacimc) [inline catalog/dags/data_refresh/indexer_worker/indexer_worker/authority.py:1]: File copied from the ingestion server without change.

Comment (stacimc) [inline catalog/dags/data_refresh/indexer_worker/indexer_worker/authority.py:1]: Copied from `ingestion_server` without changes to logic.

File: catalog/dags/data_refresh/indexer_worker/indexer_worker/db_helpers.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import logging as log
```
Comment (stacimc) [inline catalog/dags/data_refresh/indexer_worker/indexer_worker/db_helpers.py:1]: Copied from `ingestion_server` without changes to logic.

File: catalog/dags/data_refresh/indexer_worker/indexer_worker/elasticsearch_models.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
+"""
```
Comment (stacimc) [inline catalog/dags/data_refresh/indexer_worker/indexer_worker/elasticsearch_models.py:1]: Copied from `ingestion_server` without changes to logic. This provides the mappings from db record to ES document.

File: catalog/dags/data_refresh/indexer_worker/indexer_worker/es_helpers.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import logging as log
```
Comment (stacimc) [inline catalog/dags/data_refresh/indexer_worker/indexer_worker/es_helpers.py:1]: Copied from `ingestion_server` without changes to logic.

**Comment by sarayourfriend** [comment]
@stacimc I've opened a PR targeting this branch that switches Pipenv out for PDM, add a `test` recipe to the new justfile, and to run indexer worker tests in CI.

https://github.com/WordPress/openverse/pull/4497

I didn't see a separate issue for implanting CI for the new project, so I wasn't sure where they would go, if not in this PR; but if you want to keep it as a fast-follow, no worries, it doesn't need to be merged into this PR, can be on its own if you'd prefer.

**Review by sarayourfriend — APPROVED** [review]
This LGTM! On https://github.com/WordPress/openverse/pull/4497 I can run `just indexer_worker/test` and all the tests pass. On this branch, `just indexer_worker/test-local` also passes after running through the local venv installation.

I have not spent much time reviewing the code line by line given how much of it is copy/paste. That's fine by me, as well as the decision to push tests off into integration style tests in later issues :+1:

**Review by AetherUnbound — APPROVED** [review]
This looks great! I was able to test this locally and it worked as expected. A couple of comments here but nothing blocking, if anything we can make issues going forward for addressing them as non-urgent follow ups (and possibly good first issues!).

File: indexer_worker/indexer_worker/tasks.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    def get_task_status(self, task_id: str) -> dict:
+        """
+        Get the status of a single task with the given task ID.
+
+        :param task_id: the ID of the task to get the status for
+        :return: response dictionary containing all relevant info about the task
+        """
+
+        def _time_fmt(timestamp: int) -> str | None:
```
[the remark this answers is not in the document]
Comment (stacimc) [reply indexer_worker/indexer_worker/tasks.py]: Sure!

File: indexer_worker/indexer_worker/tasks.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+            :param timestamp: the timestamp to format
+            :return: the human-readable form of the timestamp
+            """
+
+            if timestamp == 0:
+                return None
+            return str(datetime.datetime.utcfromtimestamp(timestamp))
+
+        task_info = self.tasks[task_id]
```
[the remark this answers is not in the document]
Comment (stacimc) [reply indexer_worker/indexer_worker/tasks.py:41]: Of course! Sounds like a good first issue as well. I'll create an issue and link it here 👍

**Pull request merged**